src/jsx-parser.ts

```diff
@@ -0,0 +1,575 @@
+        const selfClosing = this.matchJSX('/');
+        if (selfClosing) {
+            this.expectJSX('/');
+        }
+        this.expectJSX('>');
+
+        return this.finalize(node, new JSXNode.JSXOpeningElement(name, selfClosing, attributes));
+    }
+
+    parseJSXBoundaryElement(): JSXNode.JSXOpeningElement | JSXNode.JSXClosingElement {
+        const node = this.createJSXNode();
+
+        this.expectJSX('<');
+        if (this.matchJSX('/')) {
+            this.expectJSX('/');
+            const name = this.parseJSXElementName();
+            this.expectJSX('>');
+            return this.finalize(node, new JSXNode.JSXClosingElement(name));
+        }
+
+        const name = this.parseJSXElementName();
+        const attributes = this.parseJSXAttributes();
+        const selfClosing = this.matchJSX('/');
+        if (selfClosing) {
+            this.expectJSX('/');
+        }
+        this.expectJSX('>');
+
+        return this.finalize(node, new JSXNode.JSXOpeningElement(name, selfClosing, attributes));
+    }
+
+    parseJSXEmptyExpression(): JSXNode.JSXEmptyExpression {
+        const node = this.createJSXChildNode();
+        this.collectComments();
+        this.lastMarker.index = this.scanner.index;
+        this.lastMarker.lineNumber = this.scanner.lineNumber;
+        this.lastMarker.lineStart = this.scanner.lineStart;
+        return this.finalize(node, new JSXNode.JSXEmptyExpression());
+    }
+
+    parseJSXExpression(): Node.Expression | JSXNode.JSXEmptyExpression {
+        let expression;
+
+        if (this.matchJSX('}')) {
+            expression = this.parseJSXEmptyExpression();
+        } else {
+            this.finishJSX();
+            expression = this.parseAssignmentExpression();
+            this.startJSX();
+        }
+
+        return expression;
+    }
+
+    parseJSXExpressionContainer(): JSXNode.JSXExpressionContainer {
+        const node = this.createJSXNode();
+        this.expectJSX('{');
+        const expression = this.parseJSXExpression();
+        this.expectJSX('}');
+        return this.finalize(node, new JSXNode.JSXExpressionContainer(expression));
+    }
+
+    parseJSXChildren(): JSXNode.JSXChild[] {
+        const children = [];
+
+        while (!this.scanner.eof()) {
+            const node = this.createJSXChildNode();
+            const token = this.nextJSXText();
+            if (token.start < token.end) {
+                const raw = this.getTokenRaw(token);
+                const child = this.finalize(node, new JSXNode.JSXText(token.value, raw));
+                children.push(child);
+            }
+            if (this.scanner.source[this.scanner.index] === '{') {
+                const container = this.parseJSXExpressionContainer();
+                children.push(container);
+            } else {
+                break;
+            }
+        }
+
+        return children;
+    }
+
+    parseComplexJSXElement(el: MetaJSXElement): MetaJSXElement {
```

Comment:
This is a stack-based implementation, different than the recursive variant in the old esprima-fb. It eliminates the need for the double lookahead, and yet it should still conform to the specification.